### PR TITLE
Add obsidian app

### PR DESCRIPTION
**Pull request by Botspot**
I need someone with an arm64 OS to try this and see if performance is good enough on a RPi4.

### CODE REVIEW

File: apps/Obsidian/install-64
Comment (Crilum) [inline apps/Obsidian/install-64:7]:
```suggestion
sudo mv -f /tmp/Obsidian.AppImage /opt || exit 1
sudo ln -sf /opt/Obsidian.AppImage /usr/bin/obsidian 
```

I don't think `/opt` is on `$PATH`.

**Review by Crilum — CHANGES_REQUESTED** [review]

**Comment by ryanfortner** [comment]
Works great for me! 
image

**Comment by Botspot** [comment]
> Works great for me!

You know when you've reached the ARM world when "great" means it technically launches but with many errors. 😂

**Comment by ryanfortner** [comment]
> You know when you've reached the ARM world when "great" means it technically launches but with many errors.

lol. It seems to be a GPU-related issue, although the application still runs well.

**Comment by Botspot** [comment]
@ryanfortner, do you work much on electron apps? (Would you know how to recompile this?)

**Comment by Crilum** [comment]
I have the `failed to load driver: kms_swrast` error on other Electron apps on armhf, but it doesn't affect the performance, as far as I can tell.

**Comment by Jai-JAP** [comment]
> @ryanfortner, do you work much on electron apps? (Would you know how to recompile this?)

I know about that but i am on a vacation at present and cant compile them for a while.
:\

**Comment by ryanfortner** [comment]
> @ryanfortner, do you work much on electron apps? (Would you know how to recompile this?)

Yes, I can try to recompile it for armhf.

**Comment by ryanfortner** [comment]
@Botspot I recompiled the app for armv7l: https://transfer.sh/Rd6KlE/obsidian-linux-armv7l-unpacked.zip

File: apps/Obsidian/install-64
Comment (theofficialgman) [reply apps/Obsidian/install-64:7]: the description hasn't been written. telling the user to directly launch with /opt/Obsidian.AppImage is enough....
keep in mind this is a GUI app, and users should be launching with the .destop file (from their apps list) anyway

**Comment by theofficialgman** [comment]
tried on switch, works fine (no errors)

**Comment by theofficialgman** [comment]
> @Botspot I recompiled the app for armv7l: https://transfer.sh/Rd6KlE/obsidian-linux-armv7l-unpacked.zip

very strange... even the official appimage for arm64 contains an x86_64 node module for btime 
```
binding.node: ELF 64-bit LSB shared object, x86-64, version 1 (SYSV), dynamically linked, BuildID[sha1]=e282ba155ad1c7ead0114379ddd9f3f52cd4310b, not stripped
```

**Comment by ryanfortner** [comment]
@theofficialgman did you test my armv7l binaries, the app image, or both?

**Comment by theofficialgman** [comment]
> @theofficialgman did you test my armv7l binaries, the app image, or both?

just the official arm64 appimage. I was noting the official appimage for arm64 contains an x86_64 binary in its node modules. you deleted (or didn't include) that folder in your armv7l zip.

I can't run armv7l myself. someone else will have to test

**Comment by Crilum** [comment]
@ryanfortner I just tested your armv7l binary, and I get this:
```
$ ./obsidian 
2022-05-22 20:31:57 Loading main app package /home/pi/src/obsidian-linux-armv7l-unpacked/resources/obsidian.asar
2022-05-22 20:31:57 Failed to load both app packages.
2022-05-22 20:31:57 Checking for update using Github
2022-05-22 20:31:57 Success.
2022-05-22 20:31:57 Latest version is 0.14.6
2022-05-22 20:31:57 App is up to date.
MESA-LOADER: failed to retrieve device information
MESA-LOADER: failed to retrieve device information
MESA-LOADER: failed to retrieve device information
MESA-LOADER: failed to open kms_swrast: /usr/lib/dri/kms_swrast_dri.so: cannot open shared object file: Permission denied (search paths /usr/lib/arm-linux-gnueabihf/dri:\$${ORIGIN}/dri:/usr/lib/dri)
failed to load driver: kms_swrast
MESA-LOADER: failed to open swrast: /usr/lib/dri/swrast_dri.so: cannot open shared object file: Permission denied (search paths /usr/lib/arm-linux-gnueabihf/dri:\$${ORIGIN}/dri:/usr/lib/dri)
failed to load swrast driver
```

Where it then proceeds to hang. (Unless it takes like, 5 minutes to load...)

It looked like there were some permission errors, so I tried with sudo:
```
$ sudo ./obsidian --no-sandbox
2022-05-22 20:30:25 Loading main app package /home/pi/src/obsidian-linux-armv7l-unpacked/resources/obsidian.asar
2022-05-22 20:30:25 Failed to load both app packages.
[29192:0522/143025.952743:ERROR:bus.cc(397)] Failed to connect to the bus: Could not parse server address: Unknown address type (examples of valid types are "tcp" and on UNIX "unix")
[29192:0522/143025.953023:ERROR:bus.cc(397)] Failed to connect to the bus: Could not parse server address: Unknown address type (examples of valid types are "tcp" and on UNIX "unix")
No protocol specified
No protocol specified
No protocol specified
No protocol specified
[29265:0522/143026.011865:ERROR:angle_platform_impl.cc(44)] Display.cpp:966 (initialize): ANGLE Display::initialize error 12289: Could not open the default X display.
[29265:0522/143026.012243:ERROR:gl_surface_egl.cc(808)] EGL Driver message (Critical) eglInitialize: Could not open the default X display.
[29265:0522/143026.012409:ERROR:gl_surface_egl.cc(1430)] eglInitialize OpenGL failed with error EGL_NOT_INITIALIZED, trying next display type
No protocol specified
[29265:0522/143026.013261:ERROR:angle_platform_impl.cc(44)] Display.cpp:966 (initialize): ANGLE Display::initialize error 12289: Could not open the default X display.
[29265:0522/143026.024031:ERROR:gl_surface_egl.cc(808)] EGL Driver message (Critical) eglInitialize: Could not open the default X display.
[29265:0522/143026.024333:ERROR:gl_surface_egl.cc(1430)] eglInitialize OpenGLES failed with error EGL_NOT_INITIALIZED
[29265:0522/143026.025503:ERROR:gl_ozone_egl.cc(20)] GLSurfaceEGL::InitializeOneOff failed.
2022-05-22 20:30:26 Checking for update using Github
[29265:0522/143026.044473:ERROR:viz_main_impl.cc(188)] Exiting GPU process due to errors during initialization
No protocol specified
No protocol specified
No protocol specified
No protocol specified
[29277:0522/143026.148330:ERROR:angle_platform_impl.cc(44)] DisplayVkXcb.cpp:59 (initialize): xcb_connect() failed, error 1
[29277:0522/143026.148594:ERROR:angle_platform_impl.cc(44)] Display.cpp:966 (initialize): ANGLE Display::initialize error 12289: Not initialized.
[29277:0522/143026.148831:ERROR:gl_surface_egl.cc(808)] EGL Driver message (Critical) eglInitialize: Not initialized.
[29277:0522/143026.148984:ERROR:gl_surface_egl.cc(1430)] eglInitialize SwANGLE failed with error EGL_NOT_INITIALIZED
[29277:0522/143026.149164:ERROR:gl_ozone_egl.cc(20)] GLSurfaceEGL::InitializeOneOff failed.
[29277:0522/143026.157367:ERROR:viz_main_impl.cc(188)] Exiting GPU process due to errors during initialization
No protocol specified
No protocol specified
[29291:0522/143026.238732:ERROR:gpu_init.cc(446)] Passthrough is not supported, GL is disabled, ANGLE is 
2022-05-22 20:30:26 Success.
2022-05-22 20:30:26 Latest version is 0.14.6
2022-05-22 20:30:26 App is up to date.
```
Where it hangs again.

**Comment by theofficialgman** [comment]
also just a thought, it would be a good idea to at least make a comment/thread on the forum or discord requesting official armhf support  

> If you have questions or issues about core Obsidian itself, please post them to our community: https://obsidian.md/community

it might be simple enough that they just do it

**Comment by Crilum** [comment]
@ryanfortner How did you build Obsidian? [obsidianmd/obsidian-releases](https://github.com/obsidianmd/obsidian-releases) says:
>This repo is used for hosting public releases of Obsidian, as well as our community plugins & themes directories.
>
>Obsidian is not open source software and this repo DOES NOT contain the source code of Obsidian. However, if you wish to contribute to Obsidian, you can easily do so with our extensive plugin system. A plugin guide can be found here: https://marcus.se.net/obsidian-plugin-docs/

**Comment by theofficialgman** [comment]
> @ryanfortner How did you build Obsidian? [obsidianmd/obsidian-releases](https://github.com/obsidianmd/obsidian-releases) says:

for simple electron apps that don't have any/many native modules, you can rebuild them on other architectures or other versions of electron by using the .asar files from the resources folder

**Comment by ryanfortner** [comment]
> > @ryanfortner How did you build Obsidian? [obsidianmd/obsidian-releases](https://github.com/obsidianmd/obsidian-releases) says:
> 
> 
> 
> for simple electron apps that don't have any/many native modules (aka, they are all javascript and html), you can rebuild them on other architectures or other versions of electron by using the .asar files from the resources folder

Yep, that's exactly what I did. Using the command asar extract resources/app.asar testdir, I could extract the source and then recompile it from testdir

**Comment by Crilum** [comment]
There was a thread for ARM builds: https://forum.obsidian.md/t/builds-for-arm-architectures/5942/10
It was mostly people wanting arm64, however somebody did ask for armhf, but the thread just got closed...

**Comment by Botspot** [comment]
> There was a thread for ARM builds: https://forum.obsidian.md/t/builds-for-arm-architectures/5942/10 It was mostly people wanting arm64, however somebody did ask for armhf, but the thread just got closed...

I saw that too. A common argument against armhf software support is "just use an arm64 OS on your Pi!" While annoying, those people have a point: outside the Pi zone, the larger ARM world is mostly 64-bit at this point.

@ryanfortner, how did you recompile it? I extracted the asar and tried `node app.js`, `npm install`, `npm run` - nothing worked.

**Comment by ryanfortner** [comment]
> @ryanfortner, how did you recompile it? I extracted the asar and tried `node app.js`, `npm install`, `npm run` - nothing worked.

```bash
# extract the AppImage
./Obsidian-0.14.6.AppImage --appimage-extract
cd squashfs-root/resources

# unpack app.asar
sudo npm install -g asar
asar extract app.asar tempdir
cd tempdir

# compile (you will need nodejs v14 or newer to avoid fs/promises error)
npm install
npx electron-builder --linux zip --armv7l

# then check dist folder, there will be a zip and a linux-armv7l-unpacked dir
```

**Comment by theofficialgman** [comment]
> I saw that too. A common argument against armhf software support is "just use an arm64 OS on your Pi!" While annoying, those people have a point: outside the Pi zone, the larger ARM world is mostly 64-bit at this point.

those same people are still producing x86 (i386, 32bit) windows and linux executable/binaries even though those processors are LONG gone. by contrast, armv7-A (32bit) are still in production and used in many places. this includes obsidian, they produce 32bit windows executables

yes its true that servers, game consoles, and most new SBCs use armv8 64bit compatible chips

**Comment by ryanfortner** [comment]
@Botspot I think this is good to merge for now, we can always add an `install-32` script in the future.

**Comment by theofficialgman** [comment]
needs a description and to be added to categories before merging

**Comment by Crilum** [comment]
Proposed `description` (mostly from the website):
```
Obsidian is a powerful knowledge base on top of
a local folder of plain text Markdown files. 

To run from Menu: Menu -> Accessories -> Obsidian
To run from Terminal: obsidian

Long Description:
A second brain,
for you, forever.
The human brain is non-linear: we jump from idea to idea, all the time. Your second brain should work the same.

In Obsidian, making and following connections is frictionless. Tend to your notes like a gardener; at the end of the day, sit back and marvel at your own knowledge graph.

Extensively extensible

Note-taking is incredibly personal. Tried every app, but something always irks you? You deserve better.

For those who can't custom-build a solution for themselves, Obsidian is as close as you can get, with plugins, themes, and custom CSS at your disposal. 

Notes for your Grandchildren

In our age when cloud services can shut down, get bought, or change privacy policy any day, the last thing you want is proprietary format and data lock-in.

With Obsidian, your data sits in a local folder. Never leave your life's work held hostage in the cloud again.

Vibrant Community

With more than 50,000 members in our Discord chat and 30,000 members on our forum, Obsidian has one of the largest note-taking communities.

Our community welcomes anyone who uses Obsidian or is interested in Obsidian, no matter your language, country, or field. Come join us!
```

I don't actually know how to run Obsidian (from Menu or Terminal), so double check those..

**Comment by theofficialgman** [comment]
```
/tmp/.mount_ObsidimU5vwk/obsidian: /usr/lib/aarch64-linux-gnu/libxkbcommon.so.0: version `V_1.0.0' not found (required by /tmp/.mount_ObsidimU5vwk/obsidian)
```
crap not the electron <18.2.4 again

**Comment by theofficialgman** [comment]
gosh and they don't have github issues enabled 💢

**Comment by theofficialgman** [comment]
also @Botspot yes I was testing the latest release 0.14.6 which has the electron bug... not sure why you have 0.9.17 (from way back in 2020) as part of this PR

once they update their electron, we will update the script here and add an auto updater

**Comment by theofficialgman** [comment]
ok done, now its just time to wait for them to make another release

**Comment by theofficialgman** [comment]
> > You know when you've reached the ARM world when "great" means it technically launches but with many errors.
> 
> lol. It seems to be a GPU-related issue, although the application still runs well.

@ryanfortner latest release should not have this issue (though its electron 18.0.4 so it has the buster/bionic bug)
not sure why botspot had us testing the old release 0.9.17 from 2020.

**Comment by ryanfortner** [comment]
@theofficialgman in [this comment](https://github.com/Botspot/pi-apps/pull/1850#issuecomment-1133784207), I was testing v0.14.6 appimage by accident. I just reallized that, sorry.

However the v0.9.17 appimage *does not* have the gpu error. 
image

**Comment by theofficialgman** [comment]
> @theofficialgman in [this comment](https://github.com/Botspot/pi-apps/pull/1850#issuecomment-1133784207), I was testing v0.14.6 appimage by accident. I just reallized that, sorry.
> 
> However the v0.9.17 appimage _does not_ have the gpu error.

interesting... well hopefully electron 18.0.4 (what version 0.14.6 uses) just has another gpu bug... lets hope when they updated to 18.2.4 or later that all these issues magically go away. don't you love electron?

**Comment by Crilum** [comment]
> image

@theofficialgman Did you make that? It's pretty funny :rofl:

File: apps/Obsidian/install-64
Comment (Crilum) [reply apps/Obsidian/install-64:9]: @theofficialgman I find it more convenient to just type `obsidian`, if I'm launching from Terminal..

Comment (Botspot) [reply apps/Obsidian/install-64:9]: > @theofficialgman I find it more convenient to just type `obsidian`, if I'm launching from Terminal..

Yes, but I don't want to remove the .AppImage suffix.
A symlink would be better.

Comment (Crilum) [reply apps/Obsidian/install-64:9]:
```
sudo mv -f /tmp/Obsidian.AppImage /opt || exit 1
```

Won't this line move `/tmp/Obsidian.AppImage` to the file `/opt`? Shouldn't it be:
```
sudo mv -f /tmp/Obsidian.AppImage /opt/ || exit 1
```

**Comment by theofficialgman** [comment]
@Botspot this PR is just waiting for a tester on the pi

well and if you want to add a symlink or not like suggested here: https://github.com/Botspot/pi-apps/pull/1850#discussion_r878769851

**Comment by Sussy-OS** [comment]
Testing right now..

**Comment by theofficialgman** [comment]
> Testing right now..

you will need to launch from terminal to check for errors
```
/opt/Obsidian.AppImage
```

**Comment by Sussy-OS** [comment]
@theofficialgman I imported this app, and even though I have an aarch64 OS, it is not showing up.

**Comment by theofficialgman** [comment]
> @theofficialgman I imported this app, and even though I have an aarch64 OS, it is not showing up.

you made sure you actually imported it correct? like you waiting till the popup showed Obsidian and you clicked ok, then went to install under the imported category?

**Comment by Sussy-OS** [comment]
> > @theofficialgman I imported this app, and even though I have an aarch64 OS, it is not showing up.
> 
> you made sure you actually imported it correct? like you waiting till the popup showed Obsidian and you clicked ok, then went to install under the imported category?

Oh, I forgot about the imported catagory. Thanks! Testing now.

**Comment by theofficialgman** [comment]
@Botspot no .desktop file was created in the install. I assume you meant to add one after testing?

**Comment by Sussy-OS** [comment]
Imported catagory: 
image

**Comment by theofficialgman** [comment]
> @theofficialgman my imported catagory is blank:

is your pi-apps fully updated? I would do that before reporting any issues.

**Comment by Sussy-OS** [comment]
> > @theofficialgman my imported catagory is blank:
> 
> is your pi-apps fully updated? I would do that before reporting any issues.

Yes, it is fully updated: 
image

**Comment by theofficialgman** [comment]
```
cat ~/pi-apps/data/category-overrides
```

**Comment by Sussy-OS** [comment]
> 

Hmm..
image

**Comment by Crilum** [comment]
@Sussy-OS Try installing from terminal: 
```
pi-apps install Obsidian
```

**Comment by Sussy-OS** [comment]
> @Sussy-OS Try installing from terminal:
> 
> ```
> pi-apps install Obsidian
> ```

Returns:
image
 But I have an aarch64 bit OS!
image

**Comment by theofficialgman** [comment]
```
dpkg-architecture --list
```

**Comment by Sussy-OS** [comment]
> ```
> dpkg-architecture --list
> ```

```
DEB_BUILD_ARCH=armhf
DEB_BUILD_ARCH_ABI=eabihf
DEB_BUILD_ARCH_BITS=32
DEB_BUILD_ARCH_CPU=arm
DEB_BUILD_ARCH_ENDIAN=little
DEB_BUILD_ARCH_LIBC=gnu
DEB_BUILD_ARCH_OS=linux
DEB_BUILD_GNU_CPU=arm
DEB_BUILD_GNU_SYSTEM=linux-gnueabihf
DEB_BUILD_GNU_TYPE=arm-linux-gnueabihf
DEB_BUILD_MULTIARCH=arm-linux-gnueabihf
DEB_HOST_ARCH=armhf
DEB_HOST_ARCH_ABI=eabihf
DEB_HOST_ARCH_BITS=32
DEB_HOST_ARCH_CPU=arm
DEB_HOST_ARCH_ENDIAN=little
DEB_HOST_ARCH_LIBC=gnu
DEB_HOST_ARCH_OS=linux
DEB_HOST_GNU_CPU=arm
DEB_HOST_GNU_SYSTEM=linux-gnueabihf
DEB_HOST_GNU_TYPE=arm-linux-gnueabihf
DEB_HOST_MULTIARCH=arm-linux-gnueabihf
DEB_TARGET_ARCH=armhf
DEB_TARGET_ARCH_ABI=eabihf
DEB_TARGET_ARCH_BITS=32
DEB_TARGET_ARCH_CPU=arm
DEB_TARGET_ARCH_ENDIAN=little
DEB_TARGET_ARCH_LIBC=gnu
DEB_TARGET_ARCH_OS=linux
DEB_TARGET_GNU_CPU=arm
DEB_TARGET_GNU_SYSTEM=linux-gnueabihf
DEB_TARGET_GNU_TYPE=arm-linux-gnueabihf
DEB_TARGET_MULTIARCH=arm-linux-gnueabihf
```

**Comment by Sussy-OS** [comment]
Huh.. I'm confused. I have an aarch64 bit OS, but I have a 32-bit kernel?

**Comment by theofficialgman** [comment]
you do not have an arm64 OS. you have an arm64 kernel with an armhf OS

```
    # Get OS architecture.
    case $os in
        Solaris|AIX|Haiku|IRIX|FreeMiNT)
            machine_arch=$(uname -p)
        ;;

        *)  machine_arch=$kernel_machine ;;
    esac

    [[ $os_arch == on ]] && \
        distro+=" $machine_arch"
```
^ this is the neofetch code. it reports the kernel architecture under the first line. not the OS architecture

**Comment by Sussy-OS** [comment]
> you do not have an arm64 OS. you have an arm64 kernel with an armhf OS. so the opposite of what you said above
> 
> ```
>     # Get OS architecture.
>     case $os in
>         Solaris|AIX|Haiku|IRIX|FreeMiNT)
>             machine_arch=$(uname -p)
>         ;;
> 
>         *)  machine_arch=$kernel_machine ;;
>     esac
> 
>     [[ $os_arch == on ]] && \
>         distro+=" $machine_arch"
> ```
> 
> ^ this is the neofetch code. it reports the kernel architecture under the first line. not the OS architecture

Ah, that makes more sense. I will make a Raspberry Pi OS 64-bit VM so I can test. Sounds good?

**Comment by Sussy-OS** [comment]
Going offline to insert new SD card, because vdesktop can't work with new setup window.

**Comment by Crilum** [comment]
I tested with a buster vdesktop, I had to extract the appimage, but I think that was just because I was using vdesktop:

https://user-images.githubusercontent.com/91354257/174454981-f4797afa-265d-485c-85d5-0779ba6d5258.mp4

There are still a few gpu errors, but that might also be a product of vdesktop. 🤷‍♂️

**Comment by Sussy-OS** [comment]
It seems to work, with a few errors. (Then again, I've never used Obsidian) At the end, I was trying to close Obsidian, and the terminal takes longer to respond to the close, and I *almost* used Ctrl + C, but then it stopped the program itself.

https://user-images.githubusercontent.com/98842253/174457168-726297ff-fbbb-4128-89e2-b91ee6afde58.mp4

**Comment by Crilum** [comment]
One thing I noted just now, the tab icon doesn't appear...

**Comment by Crilum** [comment]
So is the only thing holding this up the Electron version, or it there something else that needs to be done?

**Comment by theofficialgman** [comment]
I think the problem actaully is webGL is used by Obsidian and that seems to not work at all in normal chromium without all of the piOS specific chromium patches. so anything like electron, CEF, etc that uses webGL (which won't have the piOS patches/hacks) simply won't work.

I think this is just an eternal no-go until raspberry pi foundation either upstreams their patches, whatever mesa/broadcom driver changes need to get made, or google works around it themselves

**Comment by Botspot** [comment]
Other electron apps work on Pi, albeit with gpu-related errors and the usual fallback to swiftshader. @theofficialgman, is there a reason you feel that this is not usable when rendering without webgl?

**Comment by theofficialgman** [comment]
> Other electron apps work on Pi, albeit with gpu-related errors and the usual fallback to swiftshader. @theofficialgman, is there a reason you feel that this is not usable when rendering without webgl?

you tell me
we had previous issues that you noted it would hang on launch https://github.com/Botspot/pi-apps/pull/1850#issuecomment-1133972994

idk if this is still the case on the latest obsidian version (which this PR isn't on rn) on pios buster/bullseye. someone needs to test all the architectures and OSs

**Comment by Crilum** [comment]
> > Other electron apps work on Pi, albeit with gpu-related errors and the usual fallback to swiftshader. @theofficialgman, is there a reason you feel that this is not usable when rendering without webgl?
> 
> you tell me we had previous issues that you noted it would hang on launch [#1850 (comment)](https://github.com/Botspot/pi-apps/pull/1850#issuecomment-1133972994)
> 
> idk if this is still the case on the latest obsidian version (which this PR isn't on rn) on pios buster/bullseye. someone needs to test all the architectures and OSs

So it basically just needs an update and testing..

File: apps/Obsidian/description
Comment (theofficialgman) [inline apps/Obsidian/description:12]: no this isn't the case
who wrote this? if the user has appimagelauncher this is the case, yes, but that isn't installed by default nor is it really the best thing honestly since it runs as a service waiting for any and all appimage launch attempts

**Comment by Botspot** [comment]
Obsidian has now been added in commit https://github.com/Botspot/pi-apps/commit/28c5833e154edae07ecd1aaf955be6db63c018d5.

**Pull request closed without merging**
